core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -0,0 +1,388 @@
+
+// Data
+sealed abstract class ContainerData(val lastUsed: Instant)
+case class NoData() extends ContainerData(Instant.EPOCH)
+case class PreWarmedData(container: Container, kind: String) extends ContainerData(Instant.EPOCH)
+case class WarmedData(container: Container, namespace: EntityName, action: WhiskAction, override val lastUsed: Instant) extends ContainerData(lastUsed)
+
+// Events received by the actor
+case class Start(exec: Exec)
+case class Run(action: WhiskAction, msg: ActivationMessage)
+case object Remove
+
+// Events sent by the actor
+case class NeedWork(data: ContainerData)
+case object ContainerPaused
+case object ContainerRemoved
+
+/**
+ * A proxy that wraps a Container. It is used to keep track of the lifecycle
+ * of a container and to guarantee a contract between the client of the container
+ * and the container itself.
+ *
+ * The contract is as follows:
+ * 1. Only one job is to be sent to the ContainerProxy at one time. ContainerProxy
+ *    will delay all further jobs until the first job is finished for defensiveness
+ *    reasons.
+ * 2. The next job can be sent to the ContainerProxy after it indicated capacity by
+ *    sending NeedWork to its parent.
+ * 3. A Remove message can be sent at any point in time. Like multiple jobs though,
+ *    it will be delayed until the currently running job has finished.
+ *
+ * @constructor
+ * @param factory a function generating a Container
+ * @param sendActiveAck a function sending the activation via active ack
+ * @param storeActivation a function storing the activation in a persistent store
+ */
+class ContainerProxy(
+    factory: (TransactionId, String, Exec, ByteSize) => Future[Container],
+    sendActiveAck: (TransactionId, WhiskActivation) => Future[Any],
+    storeActivation: (TransactionId, WhiskActivation) => Future[Any]) extends FSM[ContainerState, ContainerData] with Stash {
+    implicit val ec = context.system.dispatcher
+
+    val unusedTimeout = 30.seconds
+    val pauseGrace = 1.second
+
+    startWith(Uninitialized, NoData())
+
+    when(Uninitialized) {
+        // pre warm a container
+        case Event(job: Start, _) =>
+            factory(
+                TransactionId.invokerWarmup,
+                ContainerProxy.containerName("prewarm", job.exec.kind),
+                job.exec,
+                256.MB)
+                .map(container => PreWarmedData(container, job.exec.kind))
+                .pipeTo(self)
+
+            goto(Starting)
+
+        // cold start
+        case Event(job: Run, _) =>
+            implicit val transid = job.msg.transid
+            factory(
+                job.msg.transid,
+                ContainerProxy.containerName(job.msg.user.namespace.name, job.action.name.name),
+                job.action.exec,
+                job.action.limits.memory.megabytes.MB)
+                .andThen {
+                    case Success(container) => self ! PreWarmedData(container, job.action.exec.kind)
```

Comment:
using `PreWarmedData` for the cold start, which is also used in the above case, of a stem cell container - should use different type? does that add anything if so other than clarity?